### ISSUE TITLE
SLADE 3.2.7 doesn't launch on Windows 7 anymore

### ISSUE THREAD

**Issue by cher-nov** [sirjuddington/SLADE#1766]
### SLADE Version

3.2.7 (32-bit)

### OS

Windows 7 Ultimate x64

### Steps to Reproduce

It wasn't mentioned anywhere in Changelog, so I've decided to submit an issue.
SLADE 3.2.7, downloaded from the official website (https://slade.mancubus.net/files/3.2.7/slade_3.2.7.7z), refuses to start with the error about some function missing in a system library.

### Crash Information

```
---------------------------
SLADE.exe - Entry Point Not Found
---------------------------
The procedure entry point CreateFile2 could not be located in the dynamic link library KERNEL32.dll. 
---------------------------
ОК   
---------------------------
```

**Comment by sirjuddington** [comment]
Hmm no idea why that would happen, what about if you try the 64bit version?

**Comment by cher-nov** [comment]
> Hmm no idea why that would happen, what about if you try the 64bit version?

Exactly the same error. This looks like a classic problem related to not specifying `_WIN32_WINNT` properly and/or using a toolset in Visual Studio that is too new.
https://stackoverflow.com/questions/25018115/libcurl-the-procedure-entrypoint-createfile2-could-not-be-located

See this pull request, for example: https://github.com/R2Northstar/NorthstarLauncher/pull/770

**Comment by p0358** [comment]
Since I happened to see this PR, may as well comment: in the linked PR it was a case of minizip-ng using that define variable to check if it's compiled at least in Win8 context and then just statically import CreateFile2 function that does not exist in Win7 yet, while older versions of said library only used that function in WinRT context (basically WinStore UWP apps): https://github.com/zlib-ng/minizip-ng/commit/9829913018955aeb5f24acfd01eb79eaaaffa02f

So defining that define to target older Windows makes it work as a workaround, whereas imho the devs of such libraries should use dynamic imports to handle functions that might not be available in older Windowses at runtime rather than compile time...

But if it's another library and it doesn't have such a way to disable such linkage, then the only hacky workaround aside from patching the library, is to create a dummy CreateFile2 function that redirects it to regular CreateFile and have the linker link that instead of importing from system libs...

**Comment by sirjuddington** [comment]
This is unlikely to be addressed by me, I don't have a Win7 system to test and I'm not particularly interested in keeping support for old OSes in general. Not to say someone else can't fix the issue and put up a PR for it

**Comment by sirjuddington** [comment]
> You don't need a "win7 system". You can run a virtual machine for testing.

Yes I know that, I mean I don't have one set up and don't plan to any time soon.

> Submitting a PR with a properly compiled binary would only temporarily solve the issue for this release version. All future releases would require the exact same process if you aren't going to make the necessary changes to your process. At that point one might as well fork the repo since they would be maintaining this part of it themselves.

Why would I suggest submitting binaries as a PR? I meant a PR with code/build config changes to fix the problem in future releases.

> Either way,, I don't understand why it took over 3 months to respond to the ticket just to say that you won't do anything about the problem despite zero indication or notice that you were dropping windows 7 support. Kind of ridiculous when we're talking about a 30+ year old game.

I work on SLADE as a hobby project in my free time, so sometimes I can go a while without responding to an issue, it's just how it is. I'd rather not drop Win7 support if possible, but as a very small minority of users are still running it, it's a pretty low priority for me to get working.

> without linux support

Not sure what you mean here, linux is fully supported.

In the meantime, there is nothing stopping you from using 3.2.6 until the issue is hopefully resolved.

**Comment by cher-nov** [comment]
@sirjuddington,

> I'd rather not drop Win7 support if possible, but as a very small minority of users are still running it, it's a pretty low priority for me to get working.

I could attempt to fix it, but please tell me what you use to build the release. VS 2022?

**Comment by Blzut3** [comment]
Probably worth noting that Microsoft plans to drop support or targeting Windows 7 with whatever they end up calling Visual C++ 18.0.  While they haven't announced the next version of Visual Studio yet that I know of, Microsoft's STL changelog is already tracking changes for it.

**Comment by sirjuddington** [comment]
> > I'd rather not drop Win7 support if possible, but as a very small minority of users are still running it, it's a pretty low priority for me to get working.
> 
> I could attempt to fix it, but please tell me what you use to build the release. VS 2022?

Yeah VS2022, there is a powershell script I use to build it for release (see https://github.com/sirjuddington/SLADE/blob/master/msvc/buildall.ps1)

**Comment by Blzut3** [comment]
> The use of visual c++ 18 just ike any version is still completely optional.

Of course, but then you're asking developers to not use new features that make maintenance easier just so a small number of people can use an old operating system.

> All I have seen is a fatpack version.

You haven't looked hard enough.  With the sense of entitlement you're giving off I'm not sure I care to say more than that.

**Comment by Blzut3** [comment]
> Well since you can apparently tell the future in what ones needs are, what "new features" would exclusively require c++ 18 that aren't available on the current version? What's "easier" is in large part a matter of opinion.

The point is that inevitably there will be some C++ feature that will require a newer toolchain to access and a decision point will come were you can either put effort into supporting Windows 7 or use the more streamlined things.  I'm into retro computing so I know to some extent there will be alternative tool chains, but for the most part people just follow what Microsoft's tools target since that's the path of least resistance.

> I'm sure you know this but clearly you only joined this discussion for other reasons than to offer a solution.

Well if "solution" is defined strictly as "Windows 7 works again" and no other outcome then I suppose that would be the case. I joined the discussion to inform that restoring Windows 7 support could be fairly temporary unless sirjuddington desires to use a soon to be frozen tool chain.  Which is useful info in many ways including being aware for...

> Courteous common practice would be to inform users of plans to drop support one version in advance.

Sure if it was an intentional change.  Based on the first reply to this ticket the drop was obviously not intentional, but it's up to someone else (i.e. cher-nov) to figure out what needs to be done.  I would guess one of the dependencies was built to assume newer Windows, which may be a new default.

> The problem I do have is the poor excuse that's being used and initial lack of communication followed by cryptic messaging that was put forth after 3 months of no response.

How is "I'm not particularly interested" cryptic?  Not knowing that "I don't have a Win7 system to test" also means VMs is fine, but after clarification insisting that not wanting to deal with a VM is an "unjustified" or "poor excuse" is in my opinion entitlement.  Especially since he did "just say so" on not wanting to support Windows 7, it just I guess wasn't the particular sequence of words you wanted to see and instead left the door open to cher-nov to figure out what went wrong if desired.

But OK, I'll take your word that you didn't mean it that way.  The bottom of the Slade downloads page has "For information about other installation options in linux, see [here](http://slade.mancubus.net/index.php?page=wiki&wikipage=Installation)."  Which informs about the Debian/Ubuntu package that I maintain.  The bottom of that wiki page in turn has a link to compilation instructions.  The AUR pkgbuild for Slade also is actively maintained.  Note however, if you have nvidia graphics you may be temporarily better off continuing with what you're doing since some users are reporting problems that I can't reproduce.  (See pinned ticket.)

**Comment by cher-nov** [comment]
> If the maintainers or [@cher-nov](https://github.com/cher-nov) don't take care of it I will likely submit the PR myself and cross the bridge regarding further support if/when the time comes.

Sure, go ahead. I honestly don't know when I'll actually be able to get around to it.

**Comment by sirjuddington** [comment]
> For all this pointless yapping spamming my notifications, I decided to fuck it and take the 3 minutes to look for the cause of problem myself:
> 
> https://github.com/AcademySoftwareFoundation/openexr/blob/3f1d975557af2a72445422396ecc8451732b4749/src/lib/OpenEXRCore/internal_win32_file_impl.h#L304-L320
> 
> Now as was already said above, you just need to pass define `_WIN32_WINNT=0x0601` and ensure that library is compiled with it, it seems the EXR file is a sub-dependency of FreeImage which is directly used in SLADE. At a glance it doesn't seem like the whole format could be easily disabled. So just use that define and it should solve that problem.

Hmm, that might be tricky to do on my end as I'm using vcpkg to build/install all the dependencies, including freeimage.

This is probably the sign I should look at dropping and replacing freeimage with something else as it is pretty much abandoned as far as I know, clashes with wxWidgets and apparently has some security vulnerabilities (#1675). I think wxWidgets itself supports all the relevant image formats, I'm just not sure it has the required palette-related functionality SLADE requires.

**Comment by Blzut3** [comment]
> Have you really stooped to arguing semantics? What would the reason be then if it literally takes 2 minutes and 4 clicks of a mouse to boot up a VM instance in a program like virtual box?

It's not arguing semantics.  You're simply refusing to accept "I don't want to" as a reason.  There doesn't have to be a reason beyond that.  If you think there does then that, in my opinion, is entitlement.

> It's 100% conflicting. Explain how it's not when first he said that he no longer wants to support an old OS followed by his next comment of not wanting to stop supporting an old OS. It doesn't get more cut and dry than this unless you're deploying some ridiculous mental gymnastics.

Again, I can see how you can arrive at this, but I assure you both posts boil down to: "I can't justify spending my time on this, but if someone else would like to go ahead."  Given p0358 and cher-nov's responses I'm fairly certain they came to the same conclusion.

I see this discussion has officially hit a loop, so if for some reason you want me to follow up further, please reach out to me by other means.  I'm fairly sure though if you disagree with me still there's probably nothing I can say to change your mind.

**Comment by sirjuddington** [comment]
Ok I think this discussion has gone on long enough and gotten completely out of hand. Offtopic replies are hidden and I've blocked @Zoran-The-Great as I've had quite enough of their incredibly rude behaviour and attacks toward me and @Blzut3.

I'm leaving this issue open as it is still something that I'd like to see fixed if possible, though this whole experience has given me even less motivation to look into it than I already had (sorry to the other users affected by this).

**Comment by sirjuddington** [comment]
[SLADE-3.2.7-no-freeimage.zip](https://github.com/user-attachments/files/19687127/SLADE-3.2.7-no-freeimage.zip)

@cher-nov can you test this build and see if it runs? This is a build of #1805 which drops FreeImage so it should no longer require CreateFile2 (I can't see it in the exe using detect-it-easy as @p0358 suggested so I think it should work)

**Comment by cher-nov** [comment]
> [@cher-nov](https://github.com/cher-nov) can you test this build and see if it runs?

It does! Thank you very much!
